File: 02_search/search_types/max_pq/max_pq.java
public class MaxPQ<Key extends Comparable<Key>>
{
  private Key[] pq;
  private int N = 0;

  public MaxPQ(int maxN)
  { pq = (Key[]) new Comparable[maxN+1]; }

  public boolean isEmpty()
  { return N == 0; }

  public int size()
  { return N; }

  public void insert(Key v)
  {
    pq[++N] = v;
    swim(N);
  }

  public Key delMax()
  {
    Key max = pq[1];
    exch(1, N--);
    pq[N+1] = null;
    sink(1);
    return max;
  }

  private boolean less(int i, int j)
  private void exch(int i, int j)
  private void swim(int k)
  private void sink(int k)
}

File: 02_search/search_types/advanced_sorts/max_pq/MaxPQ.js
'use strict';

var fs = require('fs');

class MaxPQ{
  constructor(maxN) {
    this.N = 0;
    this.pq = new Array(maxN+1);
  }
  isEmpty(){
    return this.N == 0;
  }
  size(){
    return this.N;
  }
  insert(v){
    this.pq[++this.N] = v;
    this.swim(N);
  }
  delMax(){
    let max = pq[1];
    this.exch(1, this.N--);
    this.pq[this.N + 1] = null;
    this.sink(1);
    return max;
  }
  less(i, j){
    return i < j;
  }
  exch(i, j){
    let temp = this.pq[i];
    this.pq[i] = this.pq[j];
    this.pq[j] = temp;
  }
  swim(k) { 

  }
  sink(k) {

  }
};

